cluster-provision/gocli/utils/ssh/ssh.go

```diff
@@ -0,0 +1,86 @@
+package utils
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"net"
+	"os"
+
+	"golang.org/x/crypto/ssh"
+)
+
+type SSHClient interface {
+	JumpSSH(string, bool, bool) (string, error)
+}
+
+type SSHClientImpl struct {
+	SSHPort uint16
+	NodeIdx int
+}
+
+func NewSSHClient(port uint16, idx int) *SSHClientImpl {
+	return &SSHClientImpl{
+		SSHPort: port,
+		NodeIdx: idx,
+	}
+}
+
+// Jump performs two ssh connections, one to the forwarded port by dnsmasq to the local which is the ssh port of the control plane node
+// then a hop to the designated host where the command is desired to be ran
+func (s *SSHClientImpl) JumpSSH(cmd string, root, stdOut bool) (string, error) {
+	signer, err := ssh.ParsePrivateKey([]byte(sshKey))
+	if err != nil {
+		return "", err
+	}
+	u := "vagrant"
+	if root {
+		u = "root"
+	}
+
+	config := &ssh.ClientConfig{
+		User: u,
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeys(signer),
+		},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+
+	client, err := ssh.Dial("tcp", net.JoinHostPort("127.0.0.1", fmt.Sprint(s.SSHPort)), config)
+	if err != nil {
+		return "", fmt.Errorf("Failed to connect to SSH server: %v", err)
+	}
+	defer client.Close()
+
+	conn, err := client.Dial("tcp", fmt.Sprintf("192.168.66.10%d:22", s.NodeIdx))
+	if err != nil {
+		return "", fmt.Errorf("Error establishing connection to the next hop host: %s", err)
+	}
+
+	ncc, chans, reqs, err := ssh.NewClientConn(conn, fmt.Sprintf("192.168.66.10%d:22", s.NodeIdx), config)
+	if err != nil {
+		return "", fmt.Errorf("Error creating forwarded ssh connection: %s", err)
+	}
+	jumpHost := ssh.NewClient(ncc, chans, reqs)
+	session, err := jumpHost.NewSession()
+	if err != nil {
+		log.Fatalf("Failed to create SSH session: %v", err)
+	}
+	defer session.Close()
+
+	var stderr bytes.Buffer
+	var stdout bytes.Buffer
+
+	session.Stdout = os.Stdout
+	session.Stderr = os.Stderr
+	if !stdOut {
```

Comment:
In case we have `stdOut=false` we don't return anything, right? This should be fixed and we should always return something...

Reply:
Not sure what you fully mean, in case stdOut is false we return the command output as the first variable in the function and stdErr as the second
otherwise we return stdout.String() but it doesn't matter since in any case the output you want is printed on os.Stdout. does this answer your question? if it does do you suggest any changes ?

Reply:
as of now, no
all callers don't depend on the returned value since in this pr what is being replaced is script runs. also, previously before this pr the return value was never important. all callers did `docker.Exec` with `os.Stdout`. 
however in the context of my full rewrite PR some callers do, when turning bind vfio to bash for example:
`addr, err := o.sshClient.JumpSSH(o.sshPort, o.nodeIdx, "lspci -D -d "+o.pciID, true, false)`

in this context printing to stdout isn't all that important, since this value is important in the context of the script logic only

Reply:
Thanks for the elaboration. I would suggest to drop the argument for now and introduce it later when you need.